File: y2023/clojure/day8.clj
#!/usr/bin/env bb
(require
 '[clojure.math.numeric-tower :as math]
 '[clojure.string :as str])

(defn parse [network]
  (let [parse-line (fn [l]
                     (let [pattern #"(\w\w\w) = \((\w\w\w), (\w\w\w)\)"
                           [_ start left right] (re-matches pattern l)]
                       [start [left right]]))]
    (->> (str/split-lines network)
         (map parse-line)
         (into {}))))

(defn parse-input [s]
  (let [[instructions network] (str/split s #"\n\n")]
    [instructions
     (parse network)]))

(defn move [network instruction position]
  (if-let [[left right] (get network position nil)]
    (case instruction
      \R right
      \L left)
    (throw (Exception. "unknown position"))))

(defn walk [from to instructions network]
  (loop [steps         0
         instructions' (cycle (seq instructions))
         position      from]
    (if (= position to)
      steps
      (recur (inc steps)
             (rest instructions')
             (move network (first instructions') position)))))

(defn cycle-len [from inst-len instructions network]
  (loop [steps         0
         instructions' instructions
         position      from]
    (if (and (str/ends-with? position "Z")
             (zero? (mod steps inst-len)))
      steps
      (recur (inc steps)
             (rest instructions')
             (move network (first instructions') position)))))

(defn solve-part2 [instructions network]
  (let [inst-len (count instructions)]
    (->> (keys network)
         (filter #(str/ends-with? % "A"))
         (map #(cycle-len %
                          inst-len
                          (cycle (seq instructions))
                          network))
         (reduce (fn [acc v] (math/lcm acc v))
                 1))))

(comment
  (def content (slurp "../../inputs/2023_8.txt"))
  (defn part1 [content]
    (let [[instructions network] (parse-input content)]
      (walk "AAA" "ZZZ" instructions network)))
  (defn part2 [content]
    (let [[instructions network] (parse-input content)]
      (solve-part2 instructions network)))
  (part1 content)
  (part2 content))

File: y2023/ocaml/day8.ml
open Core

type direction = Left | Right
type move = {left:string; right:string}
let parse_directions s = 
  let to_direction = function
    | 'L' -> Left
    | 'R' -> Right
    | _   -> failwith "unexpected direction"
  in
  String.to_list s
  |> List.map ~f:to_direction
  |> Sequence.cycle_list_exn

let parse_network s =
  let r = Str.regexp {|\(...\) = (\(...\), \(...\))|} in
  let parse_line line =
    if Str.string_match r line 0
    then
      (Str.matched_group 1 line,
        {left=Str.matched_group 2 line;
        right=Str.matched_group 3 line})
    else
      failwith "unexpected line"
  in
  String.split_lines s
  |> List.map ~f:parse_line
  |> Map.of_alist_exn (module String)

let parse_content s =
  match Str.split (Str.regexp "\n\n") s with
  | [instructions; network] ->
    (String.length instructions, parse_directions instructions, parse_network network)
  | _ -> failwith "unexpected content"

let cycle_length (l ,directions, network) target start =
  let move pos = function
    | Left -> (Map.find_exn network pos).left
    | Right -> (Map.find_exn network pos).right
  in
  let rec aux steps pos dirs =
    if (target pos) && (steps % l = 0)
    then steps
    else
      aux (steps + 1) (move pos (Sequence.hd_exn dirs)) (Sequence.tl_eagerly_exn dirs)
  in
  aux 0 start directions

let part1 content = 
  cycle_length (parse_content content) (String.equal "ZZZ") "AAA"
  |> Printf.printf "%d\n"

let lcm a b =
  let rec hcf a b =
    match a % b with
    | 0 -> b
    | rem -> hcf b rem
  in
  if a > b
  then a * b / (hcf a b)
  else a * b / (hcf b a)

let part2 content =
  let end_target s = phys_equal 'Z' (String.nget s 2) in
  let (_, _, network) as t = parse_content content in
  Map.keys network
  |> List.filter ~f:(fun s -> phys_equal 'A' (String.nget s 2))
  |> List.map ~f:(cycle_length t end_target)
  |> List.fold ~init:1 ~f:(fun acc v -> lcm acc v)
  |> Printf.printf "%d\n"
